### ISSUE TITLE
Convert to Maven

### ISSUE THREAD

**Issue by lsrom** [tvrenamer/tvrenamer#116]
Hi.  

Wouldn't it be better if the project was Maven based? It is pain to manage dependencies and Maven downloads all the jars on it's own so it is easier for new developers to come in. I noticed there is Maven branch already but it seems abandoned and out of date.

If it would be ok with you @vipuldelwadia @daveharris then I would make the switch to Maven and added it in pull request. I don't see any major disadvantages in this, what about you?

**Comment by vipuldelwadia** [comment]
The biggest issue with using maven is that the swt libs aren't available in a maven repo. In that maven branch I tried using a third party swt maven repo, not sure if it still works.

The other problem is that you need to create 6 different versions of the jar I.e. you need 6 executions or perhaps use maven profiles. Although we could consider creating fat jars (32 and 64 bit in one jar) - not sure if that works though!

That maven branch is quite old, but should still work - I would start by rebasing that onto master and then going from there.

If you can get it going then I would be happy to merge in a pull request :)

**Comment by lsrom** [comment]
Regarding missing swt libs in Maven repo, [these](https://mvnrepository.com/search?q=org.eclipse.swt) aren't the libraries we are looking for? :)

**Comment by vipuldelwadia** [comment]
Yay they've finally added maven repos for the libs! They're a little out of date but I'll take em :)

As for the multiple jars, the first thought I had was using maven profiles, so that could be a starting point. Another option is different assembly files. Either way it's something that needs investigating.

I've also tried using other build systems:

- [Bazel](https://www.bazel.io), in it's own branch: https://github.com/tvrenamer/tvrenamer/tree/bazel
- [Gradle](https://gradle.org) in a local branch, which I can push if there's interest :)

**Comment by johnv02139** [comment]
I really just don't have much experience with Maven.  My fork uses Ivy, which pulls things from the same central repository as Maven.  But, just like @vipuldelwadia I didn't integrate SWT.

While, again, I don't know that much about it, I do have the sense that Maven has been surpassed by gradle.  @vipuldelwadia, I would be interested in seeing that branch.

**Comment by vipuldelwadia** [comment]
Pushed gradle work to https://github.com/tvrenamer/tvrenamer/tree/gradle

**Comment by vipuldelwadia** [comment]
I also found this project which is an example of building a multiplatform swt app jar: https://github.com/jendap/multiplatform-swt

**Comment by daveharris** [comment]
I'm going to remove the "Version 0.8" milestone as there are other things we want to push out ahead of this and it doesn't seem like this is in a state ready to be used in a few days.
If this is wrong, or someone wants to push ahead with it *this week*, then change the milestone back!

**Comment by eprenamer** [comment]
I can certainly prepare a commit to use Ivy, which is the ant-compatible tool that downloads the jars from the central repository (the one used by maven).  I don't think there's anything wrong with Ant as a build tool.  The primary advantage of Maven is the automatic download and configuration of Jar files, and that can be achieved with Ivy.

**Comment by vipuldelwadia** [comment]
That seems like a good first step.

The main outcome I want from converting to maven is to have builds available on the build server (both release and snapshot) - I'm not sure if Ivy enables that?

**Comment by eprenamer** [comment]
@vipuldelwadia Sorry, not sure I understand the question.  When we add Ivy, what it will do is set up a little cache (in the user's home directory) of the specified jars, and then copy them into the work directory, and the build goes on from there.  If you run Ivy again, it checks to make sure you still have the right versions.  If the version has disappeared from your working directory, it just fetches it from the cache.

(It's possible to configure it to get the latest version of a jar, in which case, it will check the central repository and download a new version when available, but I would just specify exact versions and they'll always be up-to-date until someone specifically changes them.)

So, that's what it does.  I don't know what's the problem that currently prevents you from having builds available on the build server, so I don't know if it solves that.

**Comment by vipuldelwadia** [comment]
Ah, sorry, let me explain a little more. TravisCI (the build server we use) has functionality to store the archives built during a build and make them available for download:
https://docs.travis-ci.com/user/uploading-artifacts/

Another build system we have integrated (CircleCI) goes even further and automates all this, making the builds available to authenticated members: https://circleci.com/docs/1.0/build-artifacts/

Maven helps with this since the output of a Maven build is the artifacts generated in the `target/` folder - I thought that this would be a good excuse to convert the build system to Maven! However, I think we can probably achieve the same thing by running `ant dist` during a build and configuring the CI system to look in the right place for the zip files. (I'm not sure if Launch4j will work or even if it's supported.)

In short, converting to Ivy won't help or hinder this, I think.

**Comment by eprenamer** [comment]
Ah, understood.

However, the OP of this issue asks for easier dependency management, which Ivy does provide.  :)  This is basically at the front of my branch, so I'll submit a PR for it.   If you have concerns, I can defer it.

**Comment by eprenamer** [comment]
The SWT libraries, version 4.3, became available in the Maven central repository.  I've converted the ant project to use ivy, and also added a gradle build file.  The jar files are no longer checked in as source.

But I have not looked into the continuous integration.  So I'll leave this open until I figure out the status of that.

**Comment by vipuldelwadia** [comment]
Closing as won't do. The build stays on Gradle, and it was modernised for 1.0 to produce the native bundles.

Release: https://github.com/tvrenamer/tvrenamer/releases/tag/v1.0
